include/irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl

```diff
@@ -60,19 +60,29 @@ irr_glsl_BSDFSample irr_glsl_ggx_smith_cos_gen_sample(in irr_glsl_AnisotropicVie
     //reflect V on H to actually get L
     smpl.L = H*2.0*HdotV - interaction.isotropic.V.dir;
 
-    //==== compute probability ====
-    float a2 = _ax*_ay;
-    float lambda = irr_glsl_smith_ggx_Lambda(irr_glsl_smith_ggx_C2(interaction.isotropic.NdotV_squared, a2));
-    float G1 = 1.0 / (1.0 + lambda);
-    //here using isotropic trowbridge_reitz() instead of irr_glsl_ggx_burley_aniso()
-    smpl.probability = irr_glsl_ggx_trowbridge_reitz(a2,NdotH*NdotH) * G1 * abs(dot(interaction.isotropic.V.dir,H)) / interaction.isotropic.NdotV;
-
     return smpl;
 }
-irr_glsl_BSDFSample irr_glsl_ggx_smith_cos_gen_sample(in irr_glsl_AnisotropicViewSurfaceInteraction interaction, in uvec2 _sample, in float _ax, in float _ay)
+irr_glsl_BSDFSample irr_glsl_ggx_cos_generate(in irr_glsl_AnisotropicViewSurfaceInteraction interaction, in uvec2 _sample, in float _ax, in float _ay)
 {
     vec2 u = vec2(_sample)/float(UINT_MAX);
-    return irr_glsl_ggx_smith_cos_gen_sample(interaction, u, _ax, _ay);
+    return irr_glsl_ggx_cos_generate(interaction, u, _ax, _ay);
+}
+
+vec3 irr_glsl_ggx_cos_remainder_and_pdf(out float pdf, in irr_glsl_BSDFAnisotropicParams params, in irr_glsl_AnisotropicViewSurfaceInteraction interaction, in uvec2 u, in mat2x3 ior2, in float ax, in float ay)
+{
+	irr_glsl_BSDFSample s = irr_glsl_ggx_cos_generate(interaction, u, ax, ay);
```

Comment:
no, the API from saturday assumes that you already give the sample generated, not generate it inside the function.